### PR TITLE
[FLINK-4186] Use Flink metrics to report Kafka metrics

### PR DESCRIPTION
**Pull request by rmetzger**
We were using Flink's accumulators in the past to report the Kafka metrics.
With this change, we'll use the new metrics of Flink.

I'm now also exposing the current offset of each topic partition through the metrics.

### CODE REVIEW

**Comment by rmetzger** [comment]
This will simplify this PR: https://issues.apache.org/jira/browse/FLINK-4206

**Comment by zentol** [comment]
You can rebase on top of #2237.

File: flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Kafka08Fetcher.java [hunk 1]
```diff
@@ -161,6 +167,12 @@ public void runFetchLoop() throws Exception {
 				periodicCommitter.start();
 			}
 
+			// register offset metrics
+			if(useMetrics) {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Kafka08Fetcher.java]: missing space after if

File: flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Kafka08Fetcher.java [hunk 2]
```diff
@@ -78,7 +79,10 @@
 	private final long invalidOffsetBehavior;
 
 	/** The interval in which to automatically commit (-1 if deactivated) */
-	private final long autoCommitInterval; 
+	private final long autoCommitInterval;
+
+	/** The metric group of this operator */
+	private final MetricGroup metricGroup;
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Kafka08Fetcher.java]: the 0.8 Fetcher has a `MetricGroup` field, while the 0.9 Fetcher only creates a local variable inside run.  Let's make this more consistent.

File: flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Kafka08Fetcher.java [hunk 3]
```diff
@@ -161,6 +167,12 @@ public void runFetchLoop() throws Exception {
 				periodicCommitter.start();
 			}
 
+			// register offset metrics
+			if(useMetrics) {
+				final MetricGroup kafkaMetricGroup = metricGroup.addGroup("Kafka08Consumer");
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Kafka08Fetcher.java]: i don't think it is necessary to include the version in the group name.

File: flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java [hunk 4]
```diff
@@ -389,8 +396,36 @@ private void updateMinPunctuatedWatermark(Watermark nextWatermark) {
 				throw new RuntimeException();
 		}
 	}
-	
-	// ------------------------------------------------------------------------
+
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java]: double empty line

File: flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/metrics/KafkaMetricGetter.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.kafka.internals.metrics;
+
+import org.apache.flink.metrics.Gauge;
+import org.apache.kafka.common.Metric;
+
+
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/metrics/KafkaMetricGetter.java]: double empty line, also missing javadocs.

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/MockRuntimeContext.java [hunk 6]
```diff
@@ -161,6 +163,11 @@ public Histogram getHistogram(String name) {
 	}
 
 	@Override
+	public MetricGroup getMetricGroup() {
+		return new UnregisteredTaskMetricsGroup.DummyIOMetricGroup();
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/MockRuntimeContext.java]: You don't really have to override this method, using the TaskMetricGroup is fine.

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 7]
```diff
@@ -1235,15 +1235,104 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
...
 
-		Map<String, Object> accuResults = result.getAllAccumulatorResults();
-		// kafka 0.9 consumer: 39 results
-		if (kafkaServer.getVersion().equals("0.9")) {
-			assertTrue("Not enough accumulators from Kafka Consumer: " + accuResults.size(), accuResults.size() > 38);
+		deleteTestTopic(topic);
+	}
+
+	/**
+	 * Test metrics reporting for consumer
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java]: Can we have a test for the producer as well?

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/MockRuntimeContext.java (hunk 6 above)
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/MockRuntimeContext.java]: I was getting a null pointer exception, that's why I return dummy metrics here.

**Comment by rmetzger** [comment]
Thank you for the review @zentol.
I addressed all your comments and added metrics for committed offsets as well (I talked to the user who requested this feature and they requested this additional metric).

File: flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java [hunk 8]
```diff
@@ -65,6 +65,9 @@
 	/** The maximum number of pending non-committed checkpoints to track, to avoid memory leaks */
 	public static final int MAX_NUM_PENDING_CHECKPOINTS = 100;
 
+	/** Boolean configuration key to disable metrics tracking **/
+	public static final String KEY_DISABLE_METRICS = "flink.disable-metrics";
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java]: is this key set in `flink-conf.yaml` ?

File: flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/metrics/KafkaMetricGetter.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+package org.apache.flink.streaming.connectors.kafka.internals.metrics;
+
+import org.apache.flink.metrics.Gauge;
+
+/**
+ * Gauge for getting the current value of a Kafka metric.
+ */
+public class KafkaMetricGetter implements Gauge<Double> {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/metrics/KafkaMetricGetter.java]: calling this KafkaMetricWrapper would be more in line with the current naming scheme for these things.

File: flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java (hunk 8 above)
Comment (rmetzger) [inline flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java]: No, its passed as a property into the consumers.

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 10]
```diff
@@ -1235,15 +1236,129 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
...
+			while(true) {
+				int numPosOffsets = 0;
+				// check that offsets are correctly reported
+				for (ObjectName object : offsetMetrics) {
+					Object offset = mBeanServer.getAttribute(object, "Value");
+					if((long) offset >= 0) {
+						numPosOffsets++;
+					}
+				}
+				if(numPosOffsets == 5) {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java]: missing space after if

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 11]
```diff
@@ -1235,15 +1236,129 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
...
+					throw error.f0;
+				}
+				offsetMetrics = mBeanServer.queryNames(new ObjectName("*:key7=offsets,*"), null);
+				Thread.sleep(50);
+			}
+			Assert.assertEquals(5, offsetMetrics.size());
+			// we can't rely on the consumer to have touched all the partitions already
+			// that's why we'll wait until all five partitions have a positive offset.
+			// The test will fail if we never meet the condition
+			while(true) {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java]: space after while

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 12]
```diff
@@ -1235,15 +1236,129 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
...
+			Assert.assertTrue("No producer metrics found", producerMetrics.size() > 30);
+
+
+			LOG.info("Found all JMX metrics. Cancelling job.");
+		} finally {
+			// cancel
+			JobManagerCommunicationUtils.cancelCurrentJob(flink.getLeaderGateway(timeout));
+		}
+
+		while(jobThread.isAlive()) {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java]: space after while

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 13]
```diff
@@ -1235,15 +1236,129 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
...
+			Thread.sleep(50);
+		}
+		if(error.f0 != null) {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java]: space after if

**Comment by rmetzger** [comment]
I fixed the failing build and addressed all comments so far ;) (I was working on the code while you've reviewed it )

File: flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java [hunk 14]
```diff
@@ -1235,15 +1236,127 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
...
+			Assert.assertEquals(5, offsetMetrics.size());
+			// we can't rely on the consumer to have touched all the partitions already
+			// that's why we'll wait until all five partitions have a positive offset.
+			// The test will fail if we never meet the condition
+			while (true) {
+				int numPosOffsets = 0;
+				// check that offsets are correctly reported
+				for (ObjectName object : offsetMetrics) {
+					Object offset = mBeanServer.getAttribute(object, "Value");
+					if((long) offset >= 0) {
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java]: missing space

**Comment by zentol** [comment]
couldn't find any problem, +1 from my side.

**Comment by rmetzger** [comment]
I'm merging the change ...

**Pull request closed without merging**